crates/subspace-service/src/dsn/import_blocks/piece_validator.rs

```diff
@@ -0,0 +1,148 @@
+            records_root_cache: Mutex::new(LruCache::new(RECORDS_ROOTS_CACHE_SIZE)),
+        }
+    }
+
+    async fn get_records_root(
+        &self,
+        segment_index: SegmentIndex,
+    ) -> Result<RecordsRoot, Box<dyn Error>> {
+        // Get random peers. Some of them could be bootstrap nodes with no support for
+        // request-response protocol for records root.
+        let get_peers_result = self
+            .dsn_node
+            .get_closest_peers(PeerId::random().into())
+            .await;
+
+        match get_peers_result {
+            Ok(mut get_peers_stream) => {
+                while let Some(peer_id) = get_peers_stream.next().await {
+                    trace!(%peer_id, "get_closest_peers returned an item");
+
+                    let request_result = self
+                        .dsn_node
+                        .send_generic_request(
+                            peer_id,
+                            RootBlockRequest {
+                                segment_indexes: vec![segment_index],
+                            },
+                        )
+                        .await;
+
+                    match request_result {
+                        Ok(RootBlockResponse { root_blocks }) => {
+                            trace!(%peer_id, %segment_index, "Root block request succeeded.");
+
+                            if let Some(Some(root_block)) = root_blocks.first() {
+                                trace!(%peer_id, %segment_index, "Root block was obtained.");
+
+                                return Ok(root_block.records_root());
+                            } else {
+                                debug!(%peer_id, %segment_index, "Root block was not received.");
+                            }
+                        }
+                        Err(error) => {
+                            debug!(%peer_id, %segment_index, ?error, "Root block request failed.");
+                        }
+                    };
+                }
+                Err("No more peers for root blocks.".into())
+            }
+            Err(err) => {
+                warn!(?err, "get_closest_peers returned an error");
+
+                Err(err.into())
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl PieceValidator for RecordsRootPieceValidator {
+    async fn validate_piece(
+        &self,
+        source_peer_id: PeerId,
+        piece_index: PieceIndex,
+        piece: Piece,
+    ) -> Option<Piece> {
+        if source_peer_id != self.dsn_node.id() {
+            let segment_index: SegmentIndex = piece_index / PieceIndex::from(PIECES_IN_SEGMENT);
+
+            let maybe_records_root =
+                { self.records_root_cache.lock().get(&segment_index).copied() };
+            let records_root = match maybe_records_root {
+                Some(records_root) => records_root,
+                None => {
+                    let records_root_result = self.get_records_root(segment_index).await;
```

Comment:
Going back to our discussion today I think we need to figure out what is the canonical root block chain, download them all and only then start pulling blocks, so all root blocks should be statically known before we pull the first piece.

Reply:
Well, before that I don't see much sense merging this PR as is since at least half of this PR will be removed and the rest will be refactored.

Reply:
I'm going to add this in a separate PR. I think that we can merge it as it is because the PR adds features and the following PR will improve it. Tasks separation follows our initial plan: https://github.com/subspace/subspace/issues/1117

Reply:
This PR adds a bunch of code that I didn't review because it shouldn't be added in the first place. For instance 100% of `get_records_root` will be gone, `match` statement we're in will be gone as well, `match` below will be removed too and the whole function body will be this:
```rust

self
    .records_roots
    .get(piece_index / PieceIndex::from(PIECES_IN_SEGMENT))
    .map(|records_root| {
        is_piece_value(...).then(Some(piece))
    })
```

I really don't see the point of introducing a bunch of code just to remove almost all of it in the very next PR, it feels backwards and creates irrelevant noise in the commit history on the `main` branch.